scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetReadFn.scala

```diff
@@ -0,0 +1,237 @@
+package com.spotify.scio.parquet.read
+
+import com.spotify.scio.parquet.BeamInputFile
+import org.apache.beam.sdk.io.FileIO.ReadableFile
+import org.apache.beam.sdk.io.hadoop.SerializableConfiguration
+import org.apache.beam.sdk.io.range.OffsetRange
+import org.apache.beam.sdk.transforms.DoFn._
+import org.apache.beam.sdk.transforms._
+import org.apache.beam.sdk.transforms.splittabledofn.{OffsetRangeTracker, RestrictionTracker}
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableSet
+import org.apache.parquet.filter2.compat.FilterCompat
+import org.apache.parquet.hadoop.ParquetFileReader
+import org.apache.parquet.hadoop.api.InitContext
+import org.apache.parquet.hadoop.metadata.BlockMetaData
+import org.apache.parquet.io.{ColumnIOFactory, ParquetDecodingException, RecordReader}
+import org.apache.parquet.{HadoopReadOptions, ParquetReadOptions}
+import org.slf4j.LoggerFactory
+
+import java.util.{Set => JSet}
+import scala.jdk.CollectionConverters._
+import scala.util.Try
+
+class ParquetReadFn[T, R](
+  readSupportFactory: ReadSupportFactory[T],
+  conf: SerializableConfiguration,
+  projectionFn: SerializableFunction[T, R]
+) extends DoFn[ReadableFile, R] {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val SPLIT_LIMIT = 64000000L
+
+  private def parquetFileReader(file: ReadableFile): ParquetFileReader = {
+    val options = HadoopReadOptions.builder(conf.get()).build
+    ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+  }
+
+  @GetRestrictionCoder def getRestrictionCoder = new OffsetRange.Coder
+
+  @GetInitialRestriction
+  def getInitialRestriction(@Element file: ReadableFile): OffsetRange = {
+    val reader = parquetFileReader(file)
+
+    val offsetRangeResult = Try(new OffsetRange(0, reader.getRowGroups.size)).toEither
+    reader.close()
+
+    offsetRangeResult.fold(throw _, identity)
+  }
+
+  @NewTracker def newTracker(@Restriction restriction: OffsetRange) =
+    new OffsetRangeTracker(restriction)
+
+  @GetSize
+  def getSize(@Element file: ReadableFile, @Restriction restriction: OffsetRange): Double = {
+    val (_, size) = getRecordCountAndSize(file, restriction)
+    size
+  }
+
+  @SplitRestriction
+  def split(
+    @Restriction restriction: OffsetRange,
+    output: DoFn.OutputReceiver[OffsetRange],
+    @Element file: ReadableFile
+  ): Unit = {
+    val reader = parquetFileReader(file)
+
+    val splitResult = Try {
+      splitRowGroupsWithLimit(
+        restriction.getFrom,
+        restriction.getTo,
+        reader.getRowGroups.asScala.toSeq
+      ).foreach(output.output)
+    }.toEither
+
+    reader.close()
+
+    splitResult.fold(throw _, identity)
```

Comment:
If the goal is to guarantee `reader.close()` is always executed, I would probably use `try {} finally {}`